File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-3.kt
class Solution {
    fun removeNthFromEnd(
        head: ListNode?,
        n: Int,
    ): ListNode? {
        val dummy = ListNode(0).apply { next = head }
        var breakPtr: ListNode? = dummy
        var current: ListNode? = dummy

        for (i in 0 until n) {
            breakPtr = breakPtr?.next
        }

        while (breakPtr?.next != null) {
            breakPtr = breakPtr?.next
            current = current?.next
        }

        current?.next = current?.next?.next

        return dummy.next
    }
}

File: Data Structures & Algorithms/add-two-numbers/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        current = dummy = ListNode()
        carry = 0

        while l1 or l2:
            val1 = l1.val if l1 else 0
            val2 = l2.val if l2 else 0
            current_sum = carry + val1 + val2

            carry = current_sum // 10
            new_value = current_sum % 10

            current.next = ListNode(new_value)
            current = current.next
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None

        if carry:
            current.next = ListNode(carry)

        return dummy.next